### PR TITLE
Aparmar/upgrade azure open ai

### PR DESCRIPTION
**Pull request by ArpitaisAn0maly**
Upgrades made.
 

- Azure Open AI- 0.27 --->1.17.0
- Azure Open API- 12-01-2023 preview --> 02-01-2024
- Chat Completion API call updated
- Embedding Calls updated
- Embedding generation updated to reflect new object structure
- Tiktoken-0.4.0--->0.5.2 ( required to for 1.1 and fine tuning)
- Langchain  freeze at 0.1.16
- Langchain Model import updated to latest langchain-openai
- Langchain experimenatl-0.0.49--->0.0.52

### CODE REVIEW

**Review by georearl — APPROVED** [review]

File: app/backend/approaches/chatreadretrieveread.py [hunk 1]
```diff
@@ -170,9 +182,9 @@ async def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]
             self.chatgpt_token_limit - len(user_question)
             )
 
-        chat_completion = await openai.ChatCompletion.acreate(
-            deployment_id=self.chatgpt_deployment,
-            model=self.model_name,
+        
+        chat_completion= await self.client.chat.completions.create(
+            model=self.chatgpt_deployment,
```
Comment (dayland) [inline app/backend/approaches/chatreadretrieveread.py:187]: are we sure this is correct? The deployment name may not match model name...

File: app/backend/approaches/chatreadretrieveread.py [hunk 2]
```diff
@@ -347,9 +361,9 @@ async def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]
             #print("System Message Tokens: ", self.num_tokens_from_string(system_message, "cl100k_base"))
             #print("Few Shot Tokens: ", self.num_tokens_from_string(self.response_prompt_few_shots[0]['content'], "cl100k_base"))
             #print("Message Tokens: ", self.num_tokens_from_string(message_string, "cl100k_base"))
-            chat_completion = await openai.ChatCompletion.acreate(
-            deployment_id=self.chatgpt_deployment,
-            model=self.model_name,
+            
+            chat_completion= await self.client.chat.completions.create(
+            model=self.chatgpt_deployment,
```
Comment (dayland) [inline app/backend/approaches/chatreadretrieveread.py:366]: same as above

File: app/backend/approaches/chatreadretrieveread.py [hunk 3]
```diff
@@ -378,9 +392,9 @@ async def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]
             #print("Few Shot Tokens: ", self.num_tokens_from_string(self.response_prompt_few_shots[0]['content'], "cl100k_base"))
             #print("Message Tokens: ", self.num_tokens_from_string(message_string, "cl100k_base"))
 
-            chat_completion = await openai.ChatCompletion.acreate(
-            deployment_id=self.chatgpt_deployment,
-            model=self.model_name,
+            
+            chat_completion= await self.client.chat.completions.create(
+            model=self.chatgpt_deployment,
```
Comment (dayland) [inline app/backend/approaches/chatreadretrieveread.py:397]: same as above

File: app/backend/requirements.txt [hunk 4]
```diff
@@ -1,15 +1,14 @@
 #### Any version change made here should also be made and tested for the enrichment and function apps in /functions and /app/enrichment
-
 azure-identity==1.12.0
 Flask==2.3.2
-langchain>=0.0.157
+langchain==0.1.16
 azure-mgmt-cognitiveservices==13.5.0
-openai==0.27.0
-# azure-search-documents==11.4.0b3
+openai==1.17.0
```
Comment (dayland) [inline app/backend/requirements.txt:6]: Why are we going backwards on this?

File: app/enrichment/requirements.txt [hunk 5]
```diff
@@ -10,4 +10,4 @@ azure.search.documents==11.4.0b11
 azure-cosmos == 4.3.1
 azure-core == 1.26.4
 tenacity == 8.2.3
-openai == 0.27.0
\ No newline at end of file
+openai==1.17.0
```
Comment (dayland) [inline app/enrichment/requirements.txt:13]: why are we going back versions here?

File: app/backend/approaches/chatreadretrieveread.py (hunk 1 above)
Comment (ArpitaisAn0maly) [reply app/backend/approaches/chatreadretrieveread.py:187]: This is correct parameter name. with new version model is deployment name. 
image

File: app/backend/approaches/chatreadretrieveread.py (hunk 2 above)
Comment (ArpitaisAn0maly) [reply app/backend/approaches/chatreadretrieveread.py:366]: Replied above. It is correct.

File: app/backend/approaches/chatreadretrieveread.py (hunk 3 above)
Comment (ArpitaisAn0maly) [reply app/backend/approaches/chatreadretrieveread.py:397]: Replied. It is correct.

**Review by dayland — APPROVED** [review]

**Pull request merged**